text/3318-field-pojection.md

```diff
@@ -0,0 +1,801 @@
+`mem::transmute::<*mut Field, &mut Wrapper<Field>>`d to yield the
+projected field.
+
+### `#[inner_projecting($T)]`
+
+#### Restrictions
+This attribute cannot be added on `union`s, because it is unclear what field
+the projection would project. For example:
+```rust
+#[inner_projecting($T)]
+pub union WeirdPair<T> {
+    a: (ManuallyDrop<T>, u32),
+    b: (u32, ManuallyDrop<T>),
+}
+```
+Each field mentioning `$T` will either need to be a ZST, `#[inner_projecting]` or `$T`.
+
+#### How it works
+First each field of type `Pointer<$T>` (remember, we are projecting from
+`Wrapper<Pointer<Struct>> -> Wrapper<Pointer<Field>>`) is projected to
+`Pointer<$F>` and construct a `Wrapper<Pointer<$F>>` in place (because `Pointer<$F>`
+will have the same size as `Pointer<$T>` this will take up the same number of
+bytes, although the layout might be different).
+
+The special behavior of `Pin` is explained [below][pin-projections-section].
+
+## Interactions with other language features
+
+### Bindings
+
+Bindings are also be supported:
+```rust
+struct Foo {
+    a: usize,
+    b: u64,
+}
+
+fn process(x: &Cell<Foo>, y: &Cell<Foo>) {
+    let Foo { a: ax, b: bx } = x;
+    let Foo { a: ay, b: by } = y;
+    // ax, bx, ay and by are all &Cell;
+    ax.swap(ay);
+    bx.set(bx.get() + by.get());
+}
+```
+Enum bindings cannot be supported with the wrappers [`MaybeUninit`][maybeuninit]`<T>` and
+[`Cell`][cell]`<T>`:
+```rust
+enum FooBar {
+    Foo(usize, usize),
+    Bar(usize),
+}
+
+fn problem(foo: &Cell<FooBar>) {
+    match foo {
+        Foo(a, b) => {
+            foo.set(Bar(0));
+            // UB: access to uninhabited field!
+            let x = b.get();
+        }
+        _ => {}
+    }
+}
+```
+[`MaybeUninit`][maybeuninit]`<T>` has the problem that we cannot read the discriminant (as it might
+not be initialized).
+
+### `Deref` and `DerefMut`
+
+Field projection should have higher priority similar to how field access has a higher priority than
+`Deref`:
+```rust
+struct Foo {
+    field: usize,
+    inner: Bar,
+}
+
+struct Bar {
+    field: isize,
+}
+
+impl core::ops::Deref for Foo {
+    type Target = Bar;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+fn demo(f: &Foo) {
+    let _: usize = f.field;
+    let _: isize = (**f).field;
+}
+```
+
+Users will have to explicitly deref the expression/call `deref` explicitly.
+
+This does not introduces code breakage, because `Pin` is the only wrapper type that is `Deref`.
+And it special treatment will be explained below.
+
+## Pin projections
+[pin-projections-section]: #pin-projections
+
+Because [`Pin`][pin]`<P>` is a bit special, as it is the only Wrapper that permits access to raw
+fields when the user specifies so. It needs a mechanism to do so. This proposal has chosen two
+attributes named `#[pin]` and `#[unpin]` for this purpose. They are marker attribute and provide no
+further functionality. They will be located at `::core::pin::{pin, unpin}`.
```

Comment:
Note that `::core::pin::pin` is already an existing (unstable) macro providing safe stack pinning.